docs/blueprints/EP030.rst

```diff
@@ -0,0 +1,142 @@
+    - Italo Valcy <idasilva AT fiu DOT edu>
+    - Antonio Francisco <ajoaoff AT gmail DOT com>
+    - Rogerio Motitsuki <rogerio.motitsuki AT gmail DOT com>
+    - Jeronimo Bezerra <jbezerra AT fiu DOT edu>
+:Created: 2022-06-28
+:Kytos-Version:
+:Status: Draft
+
+****************************************
+EP030 - Link Liveness Detection via LLDP
+****************************************
+
+
+Abstract
+========
+
+This blueprint proposes a link liveness detection solution via LLDP. 
+
+
+Motivation
+==========
+
+In production networks, links can span over multiple transport technologies and carrier. Eventually, a segment of the end-to-end link can face an outage and this outage might not necessarily trigger a port down event at the endpoints. Under those circumstances, Kytos would see the endpoints (NNI) as UP, however there is no end-to-end connectivity. Link liveness is expected to detect such cases by using keepalive messages between the endpoints of a link. Whenever keepalive messages are no longer seen within a specified interval, Kytos will consider that the link is impaired. In this blueprint, our focus is on creating a solution to leverage the capabilities provided by the Link Layer Discovery Protocol protocol (IEEE 802.1AB). Other solutions and protocols are available. BFD (Bidirectional Forwarding Detection - RFC 5880) is another option that solves this problem.
+
+Requirements
+============
+
+- The Link liveness detection implementation must work bidirectionally.
+- The minimum expected liveness hello interval is 1s and the dead interval, by default, should be three times the hello interval. If three subsequent hellos aren't received, from a live/connected switch on an active/enabled interface, then the liveness status should be considered down.
+- Link liveness detection via LLDP will depend on LLDP being enabled on an interface in order to work, this is for reusing existing protocol packets and to avoid duplicating extra network packets that could become costly in terms of IO. Although link liveness will have its configuration, LLDP must be enabled, and if LLDP is disabled, link liveness will be disabled too. 
+- Link liveness detection should be configurable per interface, and the hello interval will be the same as the existing global ``POLLING_TIME`` to minimize IO and reuse the existing functionality. A new configuration option on settings ``LIVENESS_DEAD_MULTIPLIER`` will be available, this value multiplied by ``POLLING_TIME`` is the dead interval.
+- For production networks, the minimum expected hello interval that will be used is 1s. The link liveness implementation should be able to handle this frequency.
+- When liveness goes down, ``Link.status`` must be ``EntityStatus.DOWN``, that way encapsulating this to keep compatible and minimize changes on existing NApps. No new actions such as shutdown will be needed on this iteration. In order to accomplish this, a link internal metadata ``liveness_status`` will be reserved.
+
```

Comment:
@italovalcy, I'm glad you brought this point. All the features that are can influence an entity status, I believe they are supposed to work well together, and please, confirm if this is the expectation for production networks. In order for all features that can influence the protocol status of an core entity like a Link to play well together, I think they should all encapsulate and only notify topology changes based on EntityStatus, that way, only notifying when all features consider it as UP (added this information on this commit https://github.com/kytos-ng/kytos/pull/241/commits/5c43e93a2c4dcdbeeb9bef20c83d8bde7e2ca9de), so each feature only performing its single responsibility (otherwise it'd leave margin for some corner cases of incorrect notifying topology changes while other features might still consider it as protocol down or entity down, and in the future we even might have additional ones that can also add another constraints). 

Maybe we could evolve this link timer implementation to also be aligned with EntityStatus UP, just so all features work well together, originally, from what I read, it was meant to provide protection for link flap. Let me know what you think, I'm planning to map this in another issue to be aligned if you agree. 

Thanks for your input and review.

Reply:
cc'ing also @ajoaoff for his info and feedback.

Reply:
@viniarck reading the PR kytos-ng/topology#98 - which seems to implement this blueprint - it looks like the liveness status will be used to compose the Link status, along with operational status (`is_active`) and administrative status (`is_enabled`), right? 

So, for example, when the liveness is detected as UP (first hello received), it will immediately update the Link liveness metadata and will call the Link UP handler to properly generate the Link UP notification (i.e., taking into consideration the requirements from EP016 bullet 2), right?

Reply:
> @viniarck reading the PR [kytos-ng/topology#98](https://github.com/kytos-ng/topology/pull/98) - which seems to implement this blueprint - it looks like the liveness status will be used to compose the Link status, along with operational status (`is_active`) and administrative status (`is_enabled`), right?
> 
> So, for example, when the liveness is detected as UP (first hello received), it will immediately update the Link liveness metadata and will call the Link UP handler to properly generate the Link UP notification (i.e., taking into consideration the requirements from EP016 bullet 2), right?

@italovalcy, yes, in order to have these features composing well together, the `status` property will be encapsulating all the business logic, and then any notification for link up or down it would need to have a conditional checking if `link.status == EntityStatus.UP` (to notify link up) or `status != EntityStatus.UP` to notify down. However, as you highlighted, EP016 bullet 2 requirement `Timer: BOTH remote NNIS should be in the up state for more than TIMER seconds, where TIMER is defined by the network operator in a per-link basis`, this also needs to be converted as a business logic, just so, any other business logic that uses `status` can be aware of it (that's why I was bringing up that idea that NApp could register functions that could influence the status, so for instance, this conditional here https://github.com/kytos-ng/topology/blob/feature/link_liveness/main.py#L762-L763, it would also be called by the `status` property) , otherwise liveness wouldn't be aware of this, see what I mean? 

Another two practical examples:

a) [When a link goes up](https://github.com/kytos-ng/topology/blob/feature/link_liveness/main.py#L765-L767), it's checking if `link.status ==  EntityStatus.UP` to publish link status notification. Similarly, b), [when liveness goes up](https://github.com/kytos-ng/topology/blob/feature/link_liveness/main.py#L559-L562) it's also checking if `link.status ==  EntityStatus.UP`. 

Let me know if there's any point that's confusing or you're seeing any other corner case. I'll map and prototype this new requirement to allow NApps to register these functions can influence the `status` property, otherwise, without this, every time we'd have to keep more conditionals in the core like it was done in this PR #242:

```
   @property
    def status(self):
        """Return the current status of the Entity."""
        state = super().status
        if (
            state == EntityStatus.UP
            and "liveness_status" in self.metadata
            and self.metadata["liveness_status"] != "up"
        ):
            return EntityStatus.DOWN
        return state
```

Reply:
@viniarck, thank you for the detailed explanation. I don't see any other corner case for now. It all looks good to me.